src/test.cpp

```diff
@@ -1245,6 +1245,213 @@ TEST_CASE("Threshold Signatures") {
+                REQUIRE(bls::Threshold::Verify(pk, Bytes(msgHash), {sig}));
+            }
+
+            // Send participant.aggrSk sig to every other participant.
+            for (int i = 0; i < n; i++) {
+                auto& recipient = participants[i];
+                G2Element aggrSig = bls::Threshold::Sign(participant.aggrSk, Bytes(msgHash));
+                // Recv aggrSig and validates that correspond to Pa()
+                // Let's check that Pa(id) == participant.aggrPk
+                G1Element aggrPk = bls::Threshold::PublicKeyShare(finalVerifVector, Bytes(participant.id));
+                REQUIRE(aggrPk == participant.aggrSk.GetG1Element());
+                REQUIRE(bls::Threshold::Verify(aggrPk, Bytes(msgHash), {aggrSig}));
+                recipient.recvAggrSigsFrom(participant.id, aggrSig, msgHash);
+            }
+        }
+
+        // Now that everyone has everyone's sigs, let's take a participant at random and remove some sigs, then try to recover and validate SIGa(0)
+        for (int i=0; i < 10; i++) {
+            RawData random = getRandomSeed();
+            Participant p = participants[random[0] % participants.size()];
+            // Block two participants at random
+            std::vector<RawData> blockedIds;
+            blockedIds.emplace_back(participants[random[1] % participants.size()].id);
+            blockedIds.emplace_back(participants[random[2] % participants.size()].id);
+
+            // Gather sigs and ids.
+            std::vector<G2Element> aggrSigs;
+            std::vector<Bytes> ids;
+            for (const auto& recvSigShare : p.recvAggrSigs) {
+                if (std::find(blockedIds.begin(), blockedIds.end(), recvSigShare.first) != blockedIds.end()) continue;
+                ids.emplace_back(recvSigShare.first);
+                aggrSigs.emplace_back(recvSigShare.second.sig);
+            }
+
+            G2Element freeCoefficientSigs = bls::Threshold::SignatureRecover(aggrSigs, ids);
+            // This will validate against Pa(0)!
+            G1Element freeCoefficientPks = finalVerifVector[0];
+            REQUIRE(bls::Threshold::Verify(freeCoefficientPks, Bytes(msgHash), freeCoefficientSigs));
+
+            // And.. as a final check, let's craft Sa(0)
+            std::vector<PrivateKey> aggrKeys;
+            std::vector<Bytes> ids3;
+            for (const auto& participant : participants) {
+                if (std::find(blockedIds.begin(), blockedIds.end(), participant.id) != blockedIds.end()) continue;
+                aggrKeys.emplace_back(participant.aggrSk);
+                ids3.emplace_back(participant.id);
+            }
+            // Now recover the free coefficient of Sa()
+            PrivateKey finalKey = bls::Threshold::PrivateKeyRecover(aggrKeys, ids3);
+            REQUIRE(finalKey.GetG1Element() == freeCoefficientPks);
+            REQUIRE(bls::Threshold::Sign(finalKey, Bytes(msgHash)) == freeCoefficientSigs);
+        }
```

Comment:
Here, at the end, we could add a check for failed verification against the aggregated threshold signature when one of the signature shares is corrupted/invalid.

Reply:
k, sounds good. Adding it.